### DISCUSSION TITLE
Expand Record based on json value

### DISCUSSION THREAD

**Discussion by iamdineshkumar** [pocketbase/pocketbase#3516] [Q&A] [answered]
Is that possible to expand based on values in json? 
Scenario
I have two collections
1. Machineries(Holds basic details of machine)
2. Environments(Hold information of different environment and its characteristic as well as json field which has information of machineries & their performance as well as few other related metrics)
I am having a pattern of json as below

```json
{
    "commodity": "test",
    "machines": [
        {
            "id": "k55kpop7joviv6c",
            "performance": "80"
        },
        {
            "id": "5b26uz6gux1tm83",
            "performance": "70"
        }
    ]
}
```

I want to know whether machineries can be expanded through queries or is that possible to add expanded values in OnRecordsListRequest hook.

**Comment by ganigeorgiev** [comment]
For the `json` field we currently support only [`json_extract`](https://www.sqlite.org/json1.html#jex) access using dot-notation, meaning that you need to know the exact path to the json prop you want to check, eg. `yourJsonField.machines.0.performance > 50`.

> I want to know whether machineries can be expanded through queries or is that possible to add expanded values in OnRecordsListRequest hook.

I don't understand what do you mean by this.

Edit: On second read - it seems that you want expand, not filtering. Then if that's the case - No, you need a `relation` field for the expand to work.

**Reply by ganigeorgiev** [reply]
In general, if you have relation values, it will be better to use the `relation` field (either in the same collection or as junction collection) because you'll have more control over the filtering and also can make use of the auto cascade delete for example in case a "machine" is deleted.

**Comment by iamdineshkumar** [comment]
Is that possible to modify response to have machines by manually fetching details in [OnRecordsListRequest](https://pocketbase.io/docs/go-event-hooks/#onrecordslistrequest) hook by adding logic?

**Reply by ganigeorgiev** [reply, marked as answer]
You can query them manually but again in general it would be better to restructure your collection.

If you still want to manually add the machines as "expand" field, you can do something like:

```go
type machinesData struct {
    Machines []struct {
        Id          string
        Performance string
    }
}

app.OnRecordsListRequest("yourCollection").Add(func(e *core.RecordsListEvent) error {
    requestInfo := apis.RequestInfo(e.HttpContext)

    machinesCollection, err := app.Dao().FindCollectionByNameOrId("machines")
    if err != nil {
        return err
    }

    for _, r := range e.Records {
        data := machinesData{}
        if err := r.UnmarshalJSONField("yourJsonField", &data); err != nil {
            return err
        }

        // extract the machine ids
        ids := make([]string, len(data.Machines))
        for i, m := range data.Machines {
            ids[i] = m.Id
        }

        // fetch the machines with the extracted ids
        // (it is ok to perform N+1 queries in SQLite - https://sqlite.org/np1queryprob.html)
        machines, err := app.Dao().FindRecordsByIds(machinesCollection.Id, ids, func(q *dbx.SelectQuery) error {
            if requestInfo.Admin != nil {
                return nil // admins can access everything
            }

            if machinesCollection.ViewRule == nil {
                return fmt.Errorf("Only admins can view collection %q records", machinesCollection.Name)
            }
            
            // check the collection View API rule
            if *machinesCollection.ViewRule != "" {
                resolver := resolvers.NewRecordFieldResolver(app.Dao(), machinesCollection, requestInfo, true)
                expr, err := search.FilterData(*machinesCollection.ViewRule).BuildExpr(resolver)
                if err != nil {
                    return err
                }
                resolver.UpdateQuery(q)
                q.AndWhere(expr)
            }

            return nil
        })
        if err != nil {
            // return the error or just log it and leave the machines expand empty
            if app.IsDebug() {
                log.Println(err)
            }
        } else {
            // use MergeExpand in case the record has other expand fields        
            r.MergeExpand(map[string]any{"machines": machines})
        }
    }

    return nil
})
```
